neps/nep-0452.md

```diff
@@ -0,0 +1,156 @@
+type NFTData = {
+   /// the id of the token to transfer
+   token_id: string,
+   /// The valid NEAR account indicating the Non-Fungible Token contract.
+   contract_id: string
+}
+
+
+/// Data outlining Fungible Tokens that should be sent to the claiming account 
+/// (either new or existing) when a key is successfully used.
+type FTData = {
+   /// The number of tokens to transfer, wrapped in quotes and treated
+   /// like a string, although the number will be stored as an unsigned integer
+   /// with 128 bits.
+   amount: string,
+   /// The valid NEAR account indicating the Fungible Token contract.
+   contract_id: string
+}
+
+
+
+
+/****************/
+/* VIEW METHODS */
+/****************/
+
+// Returns the $yoctoNEAR amount associated with a given public key 
+// Panics if the key does not exist 
+function get_key_balance(key: PublicKey) -> string
+
+// Returns the KeyInfo associated with a given public key
+// Panics if the key does not exist
+function get_key_information(key: PublicKey) -> KeyInfo
+
+/******************/
+/* CHANGE METHODS */
+/******************/
+
+// Transfer all assets linked to the signer’s public key to an *existing* NEAR account. If the transfer fails for whatever reason, it is up to the smart contract developer to choose what should happen. For example, the contract can choose to keep the assets or send them back to the original linkdrop creator.
+// 
+// Requirements:
+// * The predecessor account *MUST* be the current contract ID.
+// * The assets being sent *MUST* be associated with the signer’s public key.
+// * The assets *MUST* be sent to the `account_id` passed in.
+//
+// Arguments:
+// * `account_id` the account that should receive the linkdrop assets.
+function claim(account_id: string) -> Promise
+
+
+
+// Creates a new NEAR account and transfers all assets linked to the signer’s public key to // the *newly created account*. If the transfer fails for whatever reason, it is up to the // smart contract developer to choose what should happen. For example, the contract can    // choose to keep the assets or send them back to the original linkdrop creator. 
+//
+// Requirements 
+// * The predecessor account *MUST* be the current contract ID.
+// * The assets being sent *MUST* be associated with the signer’s public key.
+// * The assets *MUST* be sent to the `new_account_id` passed in.
+// * The newly created account *MUST* have a new access key added to its account (either       //   full or limited access) in the same receipt that the account was created in. 
+// 
+// Arguments
+// * `new_account_id`: the valid NEAR account which is being created and should 
+//   receive the linkdrop assets
+// * `new_public_key`: the valid public key that should be used for the access key added to         //   the newly created account.
+function create_account_and_claim(new_account_id: string, new_public_key: string) -> Promise
```

Comment:
In the function `claim` and `create_account_and_claim`, does the user needs an account in advance to make these calls? Someone needs to pay for the gas.

Meta transactions could be used at this point, so the user can sign without paying the gas, but we can add here a description of that workflow, and most importantly who will be paying for the gas of that transaction.

Reply:
No, this is the whole point of linkdrops - the user signs using the private key that belongs to the contract

Reply:
I see! With that approach, however, you need to explicitly create all keys in advance on the contract, right? This will be ~~painful~~ wasteful if you plan to onboard a large number of users in a short period (on the order of thousands)

Using meta-transactions, we can create a contract that implicitly accepts other signatures.

Reply:
There are ways around this via deterministically generated keys and using entropy. Linkdrops, however, are a separate means of onboarding when compared with meta transactions. There are many pros and cons to both and I think it's a bit out of scope when talking about this standard.

Reply:
@BenKurrek 

> There are ways around this via deterministically generated keys and using entropy.

I would love to see more details about the user flow elaborated in the document.

I know that https://neardrop.io/ ([github](https://github.com/agileurbanite/contracts.near-linkdrop)) were trying to address the challenge of mass-creation of linkdrops a while ago, and they went with using factory and sub-accounts to achieve that, and it was less wasteful, but still, quite some tokens remained to be locked on the sub-accounts (created for each campaign).

Reply:
Feel free to check out [Keypom's](https://github.com/keypom/keypom#readme) implementation for some details :) 
For a quick explainer, see [here](https://docs.google.com/document/d/1x78SnZCJTnPoCDDTbnhxbjqAMPPXcsFYbKDXashRfEI/edit)